Add Python 3.4/5/6, drop 3.3/2.6

jamadden opened the pull request:
Needs a release of z3c.ptcompat to actually pass the tests.

Project gardening:

- Badges
- Rename .txt -> .rst
- Add coverage environment and coveralls
  - 100% coverage

Fixes #1.

jamadden committed:
Rename .txt -> .rst

jamadden committed:
Add 3.4/5/6, remove 3.3, 2.6. Metadata only---tests do not pass for any version.

The error is:

```
Failure in test //z3c.template/.tox/py27/lib/python2.7/site-packages/z3c/template/README.rst
Failed doctest test for README.rst
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/z3c/template/README.rst", line 0

----------------------------------------------------------------------
File "//z3c.template/.tox/py27/lib/python2.7/site-packages/z3c/template/README.rst", line 367, in README.rst
Failed example:
    print(factory(view, request)())
Exception raised:
    Traceback (most recent call last):
      File "/opt/local/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest README.rst[77]>", line 1, in <module>
        print(factory(view, request)())
      File "//z3c.template/.tox/py27/lib/python2.7/site-packages/z3c/template/template.py", line 47, in __call__
        result = self.wrapper(**kwargs)
      File "//z3c.template/.tox/py27/lib/python2.7/site-packages/zope/pagetemplate/pagetemplate.py", line 110, in __call__
        return self.pt_render(self.pt_getContext(args, kwargs))
      File "//z3c.template/.tox/py27/lib/python2.7/site-packages/zope/pagetemplate/pagetemplate.py", line 128, in pt_render
        raise PTRuntimeError(str(self._v_errors))
       - Warning: Compilation failed
       - Warning: exceptions.AttributeError: 'NoneType' object has no attribute 'rfind'
    PTRuntimeError: ['Compilation failed', "exceptions.AttributeError: 'NoneType' object has no attribute 'rfind'"]

```

Debugging in zope.pagetemplate reveals the traceback:

```
Running zope.testrunner.layer.UnitTests tests:
  Set up zope.testrunner.layer.UnitTests in 0.000 seconds.
Traceback (most recent call last):
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/zope/pagetemplate/pagetemplate.py", line 229, in _cook
    source_file, self._text, pt_engine, self.content_type)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/z3c/ptcompat/engine.py", line 52, in cook
    text, filename=source_file, keep_body=True,
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/zpt/template.py", line 192, in __init__
    super(PageTemplate, self).__init__(body, **config)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/template.py", line 116, in __init__
    self.write(body)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/template.py", line 209, in write
    self.cook(body)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/template.py", line 145, in cook
    digest = self.digest(body, names)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/zpt/template.py", line 304, in digest
    hex = super(PageTemplate, self).digest(body, names)
  File "//z3c.template/.tox/py27/lib/python2.7/site-packages/chameleon/template.py", line 246, in digest
    digest = os.path.splitext(self.filename)[0] + '-' + digest
  File "//z3c.template/.tox/py27/bin/../lib/python2.7/posixpath.py", line 98, in splitext
    return genericpath._splitext(p, sep, altsep, extsep)
  File "//z3c.template/.tox/py27/bin/../lib/python2.7/genericpath.py", line 99, in _splitext
    sepIndex = p.rfind(sep)
AttributeError: 'NoneType' object has no attribute 'rfind'

```

This appears to be an error in z3c.ptcompat.

jamadden committed:
100% test coverage. Badges. setup.cfg for universal wheels.

mgedmin commented on setup.py:
```diff
@@ -17,56 +17,66 @@
-setup (
+CHAMELEON_REQUIRES = [
+    'z3c.pt >= 2.1',
+    'z3c.ptcompat>=1.0',
```
I wish you'd pick one: spaces around `>=` or no spaces around `>=`.

jamadden replied:
In my defence, I didn't write that. [I just moved it](https://github.com/zopefoundation/z3c.template/pull/2/files#diff-2eeaed663bd0d25b7e608891384b7298L65) from lower in the file.

mgedmin replied:
Apologies for mis-assigning the blame!

mgedmin commented on setup.py:
```diff
@@ -17,56 +17,66 @@
+    ],
+    url='https://github.com/zopefoundation/z3c.template',
+    packages=find_packages('src'),
+    package_dir={'':'src'},
```
Space after `:`?

jamadden replied:
OK.

mgedmin commented on src/z3c/template/tests.py:
```diff
@@ -58,31 +49,88 @@ def setUpZPT(suite):
+        bound = z3c.template.template.BoundViewTemplate(None, None)
+        with self.assertRaisesRegexp(AttributeError,
+                                     "Can't set attribute"):
+            setattr(bound, 'im_func', 42)
```
No need for setattr, do `bound.im_func = 42`

mgedmin reviewed (approved):
LGTM, once the required dependency gets released to PyPI.

jamadden committed:
Whitespace in setup.py per review.

jamadden commented:
z3c.ptcompat 2.1.0 has been released, and these build are (going) green. Should I update the version pins or let it float?

icemac commented:
@jamadden If the new version of `z3c.ptcompat` is required to get green tests it should be required in setup.py.

icemac commented on README.rst:
```diff
@@ -0,0 +1,35 @@
+==============
+ Z3C template
```
Maybe this line should repeat the package name.

jamadden replied:
I didn't adjust the text, just the over/underlines to make them consistent and support the extra level of indentation I needed. I can alter this text.

icemac replied:
`Z3C` does not exist as an acronym anywhere else but maybe the English language enforces it to write this in upper case here because it is a headline and `Z3c` would look even more ugly.

mgedmin replied:
I've always understood Z3C to be an abbreviation for Zope 3 Community, and yeah, abbreviations are upper-cased in English.

icemac commented on setup.py:
```diff
@@ -17,56 +17,66 @@
+    namespace_packages=['z3c'],
+    extras_require={
+        'test': TESTS_REQUIRE,
+        'chameleon': CHAMELEON_REQUIRES,
```
Maybe it should be called `CHAMELEON_REQUIRE` so it looks more like `TESTS_REQUIRE`.

jamadden replied:
I know that looks awkward, but that's a subject/verb agreement issue in English. Plural subjects, like "tests" (which terminology we inherit from the `setup` keyword) need a plural verb...which happens to be "require" (without an 's'), while singular subjects, like 'chameleon' need a singular verb, which happens to be "requires" (with the 's'). So "tests requires" would be as wrong as "chameleon  require". 

Go figure.

icemac replied:
@jamadden Okay, you found out what you already knew: I'm not a native speaker.

`CHAMELEONS_REQUIRE` FTW ;)

jamadden replied:
@icemac Honestly, I had to double check your profile to be sure. I really hope I didn't sound condescending. Your English is much better than my German ever was, even when I was actively studying it.

icemac replied:
@jamadden Thank you. Any help to improve my English is appreciated.

jamadden commented:
@icemac I can adjust the version pins. My thought was that the update to z3c.ptcompat is only needed if you use Chameleon >= 3.0; an older version should still work if you use an older Chameleon. Plone uses [Chameleon 2.25 (*still*)](https://github.com/plone/buildout.coredev/blob/5.1/versions.cfg#L131), and it also uses [z3c.ptcompat/template 2.0](https://github.com/plone/buildout.coredev/blob/5.1/versions.cfg#L88). I just didn't want to make a staggered update harder for them.

icemac reviewed (approved):
LGTM besides to little questions.

icemac commented:
@jamadden Maybe we should require Chameleon >= 3.0 here, too. Even if this might require an increased major version number.

jamadden commented:
@icemac This package doesn't explicitly depend on Chameleon at the moment (despite the name of the extra), just z3c.pt[compat]. I can add one if desired.

It doesn't really matter to me, I know the compatibility issues in my own code and always use a latest Chameleon anyway. These projects all test against that too. I was just trying to ease things for other projects (but I'm an outsider there too, so this is only guess work). I guess the default should be "explicit is better than implicit" which would argue for higher version pins and a dep on Chameleon.

icemac commented:
@jamadden +1 for higher version pins and a dep on Chameleon. This fails early if someone is using a not compatible version combination.

jamadden commented:
I've updated the deps and adjusted the text. I wound up pushing those to #4, the approved-PR that follows on from this. Rather than deal with cleaning that up, I will merge these two in sequence.

Thank you!
